### ISSUE TITLE
file_exists() requires an extension

### ISSUE THREAD

**Issue by ned21** [quattor/pan#97]
```
"/test" = {
  if( file_exists("dir/struct") ) }
     create("dir/struct");
  } else {
    null;
  };
};
```

If dir/struct.pan or dir/struct.tpl exists then I would expect to find the value in /test. However file_exists is always false so I expect it's not checking the extensions automatically.

Workaround is to use exists() instead which handles the suffix correctly.

```
"/test" = {
  if( exists("dir/struct") ) }
     create("dir/struct");
  } else {
    null;
  };
};
```

**Comment by stdweird** [comment]
you mean `template_exists`? `file_exists` is for people who like to ship existing files as-is with filecopy, so the argument is the file

**Comment by ned21** [comment]
Agree a ```template_exists()``` function would be a better way to solve this.

**Comment by jrha** [comment]
Can `if_exists()` be used to achieve what you want?

**Comment by jrha** [comment]
Not a bug, behaves as designed (and documented) but there is missing functionality (add a `template_exists()` function)

**Comment by jouvin** [comment]
I have the feeling that it exists... Have you checked it the `exists()` function with a relative name (relative to the loadpath) is not doing the job?

**Comment by jouvin** [comment]
I confirm. From the `exists()` synopsis:
{{{
boolean exists(var);
variable_expression var;
boolean exists(path);
string path;
boolean exists(tpl);
string tpl;
}}}

**Comment by jouvin** [comment]
About `if_exists()`, it is dedicated to templates but has a different use case. It is mainly intended to be used in `include` to achieve a conditional include based on the template existence. It has the nice behaviour of returning either the template name to be included (the string passed as an argument in fact) or undef when the template doesn't exist. This is basically a nice shortcut for:
```
if ( exists('tpl') 'tpl' else undef; 
```

**Comment by jrha** [comment]
`exists()` is dangerous because it tells you if anything with that name exists, be it a variable, path in the configuration or a template, it's horrible and not to be recommended.

**Comment by jrha** [comment]
WRT `if_exists()`, does this get you what you need for now?
```pan
"/test" = {
  if (is_defined(if_exists("dir/struct"))) }
     create("dir/struct");
  } else {
    null;
  };
};
```

**Comment by jrha** [comment]
```pan
function template_exists = is_defined(if_exists(ARGV[0]));
```

**Comment by jouvin** [comment]
I disagree as every kind of object is non ambiguous due to the diference in permitted/starting characters. It's really safe.

**Comment by jrha** [comment]
Huh? I have clearly misunderstood something. I really hate overloaded functions :unamused:.

**Comment by jouvin** [comment]
The point is that:
* a variable name cannot contain any `/`.
* a configuration path must start with a `/`
* a template name must be relative (not start with a `/') but can contain (this is almost always the case) a `/ ` denoting the Pan namespace. Note that the template name is not the file path name but really the Pan namespaced name.
There is a potential ambiguity only between a variable name and a template name not containing a namespace, something that probably doesn't exist these days...

**Comment by ned21** [comment]
Is this an example of where exists() may do the wrong thing?

```
"/test" = {
  structpath = "structures/" + value('/some/path');
  if( exists(structpath) ) }
     create(structpath);
  } else {
    null;
  };
};
```

if_exists is an interesting workaround I will try out.

**Comment by jouvin** [comment]
I don't see any reason why the example you gave would not work... except if the loadpath is incorrect and if `structures/...` cannot be located using it. You may want to add a `debug()` message to print the loadpath...
`if_exists()` can certainly work but `create()` conversely to `include` doesn't accept an `undef` argument and you'll have to do something like:
```
if ( is_defined(if_exists(structpath)) ) {
   create(structpath);
};
```

or probably better:
```
"/test" = {
  structpath = if_exists("structures/" + value('/some/path'));
  if( is_defined(structpath) ) }
     create(structpath);
  } else {
    null;
  };
};
```

**Comment by jouvin** [comment]
I answered too quickly. Yes, the example provided tests the existence of variable `structpath`, not of the template! The documentation has an example about it:
```
Note that if the argument is a variable expression, only the existence of the variable
is checked. For example, the following code will always leave r with a value of
true.
    v = '/some/absolute/path';
    r = exists(v);
If you want to test the path, remove the ambiguity by using a construct like the
following:
    v = '/some/absolute/path';
    r = exists(v+'');
The value of r in this case will be true if /some/absolute/path exists or
false otherwise.
``` 
It's true that using `if_exists()` is probably less error prone...

**Comment by stdweird** [comment]
@ned21 can this be closed?

**Comment by ned21** [comment]
The problem still exists. There's a workaround but ```if (is_defined(if_exists("dir/struct"))) {``` is pretty ugly and unintuitive.
